File: src/main/java/com/hedvig/botService/chat/ConversationFactoryImpl.java
package com.hedvig.botService.chat;

import com.hedvig.botService.serviceIntegration.claimsService.ClaimsService;
import com.hedvig.botService.serviceIntegration.memberService.MemberService;
import com.hedvig.botService.serviceIntegration.productPricing.ProductPricingService;
import com.hedvig.botService.services.LocalizationService;
import com.hedvig.botService.services.triggerService.TriggerService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Component;

@Component
public class ConversationFactoryImpl implements ConversationFactory {
  private final Logger log = LoggerFactory.getLogger(ConversationFactoryImpl.class);
  private final MemberService memberService;
  private final LocalizationService localizationService;
  private final ProductPricingService productPricingService;
  private final TriggerService triggerService;
  private final ApplicationEventPublisher eventPublisher;
  private final ClaimsService claimsService;

  private final StatusBuilder statusBuilder;
  private Integer queuePos;
  private String appleUserEmail;
  private String appleUserPwd;

  public ConversationFactoryImpl(
    MemberService memberService,
    ProductPricingService productPricingService,
    TriggerService triggerService,
    ApplicationEventPublisher eventPublisher,
    ClaimsService claimsService,
    StatusBuilder statusBuilder,
    LocalizationService localizationService,
    @Value("${hedvig.waitlist.length}") Integer queuePos,
    @Value("${hedvig.appleUser.email}") String appleUserEmail,
    @Value("${hedvig.appleUser.password}") String appleUserPwd) {
    this.memberService = memberService;
    this.productPricingService = productPricingService;
    this.triggerService = triggerService;

    this.eventPublisher = eventPublisher;
    this.claimsService = claimsService;
    this.statusBuilder = statusBuilder;
    this.localizationService = localizationService;
    this.queuePos = queuePos;
    this.appleUserEmail = appleUserEmail;
    this.appleUserPwd = appleUserPwd;
  }

  @Override
  public Conversation createConversation(Class<?> conversationClass) {

    if (conversationClass.equals(CharityConversation.class)) {
      return new CharityConversation(this, memberService, productPricingService, eventPublisher, localizationService);
    }

    if (conversationClass.equals(ClaimsConversation.class)) {
      return new ClaimsConversation(eventPublisher, claimsService, productPricingService, this, memberService, localizationService);
    }

    if (conversationClass.equals(MainConversation.class)) {
      return new MainConversation(this, eventPublisher, localizationService);
    }

    if (conversationClass.equals(OnboardingConversationDevi.class)) {
      final OnboardingConversationDevi onboardingConversationDevi =
          new OnboardingConversationDevi(
              memberService, productPricingService, eventPublisher, this, localizationService, appleUserEmail,appleUserPwd);
      onboardingConversationDevi.setQueuePos(queuePos);
      return onboardingConversationDevi;
    }

    if (conversationClass.equals(TrustlyConversation.class)) {
      return new TrustlyConversation(triggerService, memberService, eventPublisher, localizationService);
    }

    if (conversationClass.equals(FreeChatConversation.class)) {
      return new FreeChatConversation(statusBuilder, eventPublisher, productPricingService, localizationService);
    }

    if (conversationClass.equals(CallMeConversation.class)) {
      return new CallMeConversation(eventPublisher, localizationService);
    }

    if(conversationClass.equals(MemberSourceConversation.class)) {
      return new MemberSourceConversation(eventPublisher, localizationService);
    }

    return null;
  }

  @Override
  public Conversation createConversation(String conversationClassName) {
    try {
      Class<?> concreteClass = Class.forName(conversationClassName);
      return createConversation(concreteClass);
    } catch (ClassNotFoundException ex) {
      log.error("Could not create conversation for classname: {}", conversationClassName, ex);
    }

    return null;
  }
}

File: src/main/java/com/hedvig/botService/chat/ConversationFactoryImpl.kt
package com.hedvig.botService.chat

import com.google.i18n.phonenumbers.PhoneNumberUtil
import com.hedvig.botService.chat.house.HouseOnboardingConversation
import com.hedvig.botService.enteties.UserContext
import com.hedvig.botService.serviceIntegration.claimsService.ClaimsService
import com.hedvig.botService.serviceIntegration.lookupService.LookupService
import com.hedvig.botService.serviceIntegration.memberService.MemberService
import com.hedvig.botService.serviceIntegration.productPricing.ProductPricingService
import com.hedvig.botService.serviceIntegration.underwriter.Underwriter
import com.hedvig.botService.services.triggerService.TriggerService
import com.hedvig.libs.translations.Translations
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Component

@Component
class ConversationFactoryImpl(
    private val memberService: MemberService,
    private val lookupService: LookupService,
    private val underwriter: Underwriter,
    private val productPricingService: ProductPricingService,
    private val triggerService: TriggerService,
    private val eventPublisher: ApplicationEventPublisher,
    private val claimsService: ClaimsService,
    private val statusBuilder: StatusBuilder,
    private val translations: Translations,
    @Value("\${hedvig.waitlist.length}") private val queuePos: Int?,
    @Value("\${hedvig.appleUser.email}") private val appleUserEmail: String,
    @Value("\${hedvig.appleUser.password}") private val appleUserPwd: String,
    private val phoneNumberUtil: PhoneNumberUtil
) : ConversationFactory {
    private val log = LoggerFactory.getLogger(ConversationFactoryImpl::class.java)

    override fun createConversation(conversationClass: Class<*>, userContext: UserContext): Conversation {

        if (conversationClass == CharityConversation::class.java) {
            return CharityConversation(this, memberService, productPricingService, eventPublisher, translations, userContext)
        }

        if (conversationClass == ClaimsConversation::class.java) {
            return ClaimsConversation(
                eventPublisher,
                claimsService,
                productPricingService,
                this,
                memberService,
                translations,
                userContext
            )
        }

        if (conversationClass == MainConversation::class.java) {
            return MainConversation(this, eventPublisher, translations, userContext)
        }

        if (conversationClass == OnboardingConversationDevi::class.java) {
            val onboardingConversationDevi = OnboardingConversationDevi(
                memberService,
                underwriter,
                eventPublisher,
                this,
                translations,
                appleUserEmail,
                appleUserPwd,
                phoneNumberUtil,
                userContext
            )
            onboardingConversationDevi.queuePos = queuePos
            return onboardingConversationDevi
        }

        if (conversationClass == HouseOnboardingConversation::class.java) {
            val houseOnboardingConversation = HouseOnboardingConversation(
                memberService, lookupService, eventPublisher, this, translations, userContext
            )
            houseOnboardingConversation.queuePos = queuePos
            return houseOnboardingConversation
        }

        if (conversationClass == TrustlyConversation::class.java) {
            return TrustlyConversation(triggerService, eventPublisher, translations, userContext)
        }

        if (conversationClass == FreeChatConversation::class.java) {
            return FreeChatConversation(statusBuilder, eventPublisher, productPricingService, translations, userContext)
        }

        if (conversationClass == CallMeConversation::class.java) {
            return CallMeConversation(eventPublisher, translations, userContext)
        }

        return if (conversationClass == MemberSourceConversation::class.java) {
            MemberSourceConversation(eventPublisher, translations, userContext)
        } else throw RuntimeException("Failed to create conversation")

    }

    override fun createConversation(conversationClassName: String, userContext: UserContext): Conversation {
        try {
            val concreteClass = Class.forName(conversationClassName)
            return createConversation(concreteClass, userContext)
        } catch (ex: ClassNotFoundException) {
            log.error("Could not create conversation for classname: {}", conversationClassName, ex)
        }

        throw RuntimeException("Failed to create conversation")
    }
}
